### PR TITLE
Push package querying to another thread

### LINKED ISSUE #20101: PackageInstallerService.ProcessProjectChange is sending non-fatal watsons causing UI delays
Investigating UI delays in https://developercommunity.visualstudio.com/content/problem/66362/visual-studio-2017-keeps-freezing-on-me.html, I see a bunch are proceeded immediately by the PackageInstallerService sending on non-fatal watsons on UI thread for the following:

```
System.ArgumentException: An item with the same key has already been added.
 at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
 at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
 at System.Collections.Generic.Dictionary`2.System.Collections.Generic.ICollection<System.Collections.Generic.KeyValuePair<TKey,TValue>>.Add(KeyValuePair`2 keyValuePair)
 at Roslyn.Utilities.ICollectionExtensions.AddRange[T](ICollection`1 collection, IEnumerable`1 values)
 at Microsoft.VisualStudio.LanguageServices.Packaging.PackageInstallerService.ProcessProjectChange(Solution solution, ProjectId projectId)
 --- End of stack trace from previous location where exception was thrown ---
 at 
```

I'm guessing this line is failing: 
https://github.com/dotnet/roslyn/blob/cdde5bd4d36d06b7f8ebe434e82eb5e35b4e1277/src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs#L462.

### PR DESCRIPTION
**Pull request by ryzngard**
Once the DteProject is retrieved, the work to get installed packages can be done off the UI thread. 

This will help reduce hanging from long running NuGet  commands. 

Fixes #20101

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
> Once the DteProject is retrieved, the work to get installed packages can be done off the UI thread.

How? The use of the EnvDTE.Project will still marshal back via COM. Won't that cause a deadlock?

**Comment by ryzngard** [comment]
>The use of the EnvDTE.Project will still marshal back via COM 

That's true, but we're moving the work to even more isolation. If it _still_ is too long and delaying the UI, we have data to back up saying it takes too long and we need to fix that. Moving it off the UI thread just makes sure we're doing a best effort to not have long standing work on the UI thread. 

We avoid lock here because nothing is waiting on this task. Probably worthy of a comment so nobody does that in the future, since it's not obvious.

**Comment by jasonmalinowski** [comment]
I guess to me I don't see the benefit of this change. We're "moving it off the UI thread", but is the expensive bit still going to marshal right back anyways? Has the NuGet team confirmed their stuff *can* be called off the UI thread?

**Comment by jasonmalinowski** [comment]
Ah, looking at the original bug I see a bit more subtle bit: the problem is that an exception is being thrown, and we're spending time (I guess) reporting that crash. So even if this is on a background thread, we're still going to block VS while the non-fatal reporting is happening.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 1]
```diff
@@ -469,14 +473,17 @@ private void ProcessProjectChange(Solution solution, ProjectId projectId)
...
             // Project was changed in some way.  Let's go find the set of installed packages for it.
-            var dteProject = _workspace.TryGetDTEProject(projectId);
+            return _workspace.TryGetDTEProject(projectId);
+        }
+
+        private void ProcessProjectChange(EnvDTE.Project dteProject, ProjectId projectId)
+        {
             if (dteProject == null)
             {
-                // Don't have a DTE project for this project ID.  not something we can query NuGet for.
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: why was this comment removed?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 2]
```diff
@@ -469,14 +473,17 @@ private void ProcessProjectChange(Solution solution, ProjectId projectId)
...
             }
 
             // Project was changed in some way.  Let's go find the set of installed packages for it.
-            var dteProject = _workspace.TryGetDTEProject(projectId);
+            return _workspace.TryGetDTEProject(projectId);
+        }
+
+        private void ProcessProjectChange(EnvDTE.Project dteProject, ProjectId projectId)
+        {
             if (dteProject == null)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: 1. this should have a threading assert.
2. below, you call into _packageService.GetInstalledPackages.  how do you know that is safe to call on this thread?

**Comment by CyrusNajmabadi** [comment]
i'm not sure how, at all, this is safe.  from talking to nuget before, this was a service that would marshal back to the UI thread.  so i don't see how this helps.

**Comment by CyrusNajmabadi** [comment]
> This will help reduce hanging from long running NuGet commands.

how?  the long running command still happens on the ui thread.

**Comment by ryzngard** [comment]
> If we're getting a crash here, why not just fix the actual crash instead of trying to mitigate the watson impact? If someone is hitting this, we should be able to use the actual reported watsons to find the issue and fix it.

Only about 20% of the UI hangs include Watson in the callstack, so it's not just NFW reporting. 

I'm following up with the NuGet team on the thread expectations and safety of the API

**Comment by CyrusNajmabadi** [comment]
> Only about 20% of the UI hangs include Watson in the callstack, so it's not just NFW reporting.

What's the other 80%?

**Comment by ryzngard** [comment]
The other 80% have `ProcessProjectChange` at the top of the stack with no exception

**Comment by ryzngard** [comment]
I'm evaluating other mitigations that are less risky. I don't think this is the right fix, especially for P3

**Pull request closed without merging**